### PR TITLE
feat(ec2): ipv6 support for new nics

### PR DESCRIPTION
**Pull request by aciba90**
## do not squash

https://warthogs.atlassian.net/browse/SC-1754

### CODE REVIEW

File: pycloudlib/ec2/instance.py
Comment (CalvoM) [inline pycloudlib/ec2/instance.py:126]: I think you meant `ipv4_public_ip_count`.

File: pycloudlib/ec2/instance.py
Comment (CalvoM) [inline pycloudlib/ec2/instance.py:214]: Why have you opted for this syntax?

**Review by CalvoM — COMMENTED** [review]
This is good. Could you please provide a sample script to test the functionality added?

File: pycloudlib/ec2/instance.py
Comment (aciba90) [reply pycloudlib/ec2/instance.py:214]: I am creating a copy of that list to iterate over because I want to remove the created interfaces from the list while being removed from this loop or manually.

See:

https://github.com/canonical/pycloudlib/pull/355/files#diff-0c9a2ff75eae3b564b5813cf4f5d5ea8cb15fbbf3b680b780fcb102503dc7e5cR550

File: pycloudlib/ec2/instance.py
Comment (aciba90) [reply pycloudlib/ec2/instance.py]: Good catch, thanks. Updated.

**Comment by aciba90** [comment]
> This is good. Could you please provide a sample script to test the functionality added?

Thanks for the review. As for testing, one can execute the tests modified in https://github.com/canonical/cloud-init/pull/5102/. If not, to test this, one can:

1. Create an ec2 instance.
2. Add new interface with different values for the new parameters.
3. Remove the interface / instance.

File: pycloudlib/ec2/instance.py
Comment (CalvoM) [reply pycloudlib/ec2/instance.py:214]: Okay I had not seen that.

File: pycloudlib/ec2/instance.py
Comment (blackboxsw) [inline pycloudlib/ec2/instance.py:501]: By placing these two operations in the same try block do we risk failing to attempt to release_address any time disassociate_address errors for each association? Is it worth separating these cases so we at least attempt both operations even if `disassociate_address` fails?

File: pycloudlib/ec2/cloud.py
Comment (blackboxsw) [inline pycloudlib/ec2/cloud.py:364]:
```suggestion
            enable_ipv6: indicates if ipv6 IMDS support must be added to the
```

File: pycloudlib/ec2/instance.py
Comment (blackboxsw) [inline pycloudlib/ec2/instance.py:498]: Do we even need to separately call disassociate_address? [Botocore seems to indicate that release_address will automatically disassociate](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2/client/release_address.html#release-address)

**Review by blackboxsw — COMMENTED** [review]
LGTM. Minor thought on potentially either:
- not running disassociate_address in the first place because release_address may encapsulate that operation
- or a suggestion to still attempt to run release_address even if disassociate_address fails because of the first point above

File: pycloudlib/ec2/instance.py
Comment (aciba90) [reply pycloudlib/ec2/instance.py:498]: That's only true in the default VPC:

> [Default VPC] Releasing an Elastic IP address automatically disassociates it from any instance that it’s associated with. To disassociate an Elastic IP address without releasing it, use DisassociateAddress.

> [Nondefault VPC] You must use DisassociateAddress to disassociate the Elastic IP address before you can release it. Otherwise, Amazon EC2 returns an error ( InvalidIPAddress.InUse).

File: pycloudlib/ec2/instance.py
Comment (aciba90) [reply pycloudlib/ec2/instance.py:501]: If `disassocite_address` fails, it is very likely that `release_address` is going to fail, as in non-default VPCs, address must be disassociated as a precondition before `release_address` can be performed.

File: pycloudlib/ec2/instance.py
Comment (aciba90) [reply pycloudlib/ec2/instance.py]: Done.

File: pycloudlib/ec2/cloud.py
Comment (aciba90) [reply pycloudlib/ec2/cloud.py]: Done.

File: pycloudlib/ec2/instance.py
Comment (aciba90) [reply pycloudlib/ec2/instance.py]: Double try-except added, thanks.

File: pycloudlib/ec2/instance.py
Comment (blackboxsw) [reply pycloudlib/ec2/instance.py:498]: Thanks for this @aciba90

File: pycloudlib/ec2/instance.py
Comment (blackboxsw) [inline pycloudlib/ec2/instance.py:130]: if ipv4_address_count == 1, then we aren't going to call allocate_address/associate_address below. In checking the launched instances,I'm already seeing two private ips before any additional allocation based on ipv4_address_count > 1. Do we already expect that initial IPv4 private address to be allocated and associated already without needing to call allocate_address or associate_address within the for i in range(ipv4_address_count) below? ? 

Also, if ipv4_address_count  == 2, we'll call allocate_address/associate_address twice because our range below is for i in range(ipv4_address_count). Does this actually add 2 extra private IPs to this interface even though it may already have 1 private IP associated with it? Should that range actually be `range(ipv4_address_count -1)` instead?

Also, in testing this PR with the corresponding draft PR in cloud-init/ec2-ipv6 on the integration test  tests/integration_tests/test_networking.py::test_ec2_multi_nic_reboot. I'm getting botocore errors that we are allocating too many addresses. I'm wondering if that's just on my account, or due to the approach we are taking to continue to allocate more addresses than are typically allowed for an instance 
```
E           botocore.exceptions.ClientError: An error occurred (AddressLimitExceeded) when calling the AllocateAddress operation: The maximum number of addresses has been reached.
```

File: pycloudlib/ec2/instance.py
Comment (blackboxsw) [inline pycloudlib/ec2/instance.py:151]: allocate_address allocates an elastic IP address. Shouldn't we only be doing this below in the `if associate_public_ip` scope? This is probably the reason I was seeing limits on public address allocation when running integration tests.

Don't we instead want `assign_private_ip_addresses` here  instead of the allocate_address/associate_address?

File: pycloudlib/ec2/instance.py
Comment (aciba90) [reply pycloudlib/ec2/instance.py:127]: I messed up, by copy-pasting, `ipv4_address_count` and `ipv4_public_count`. Expectations are that the default NIC (the one associated with the instance at launch) will have a private and public ipv4 by default so pycloudlib does not have to create them. I have fixed it, added comments and improved the documentation to reflect that.

File: pycloudlib/ec2/instance.py
Comment (aciba90) [reply pycloudlib/ec2/instance.py]: Yes, iterating over `range(1, ipv4_public_ip_count)` skipping the first NIC as it comes with a public ipv4 by default.

`associate_public_ip` is for a per-NIC global ip, as opposed to adding public ips attached to specific addreses.

**Comment by aciba90** [comment]
Thanks, @blackboxsw, for the review and catching up the leaking resources! I have fixed it and added docs and comments about.

File: VERSION
Comment (blackboxsw) [inline VERSION:1]:
```suggestion
1!6.7.0
```

File: pycloudlib/ec2/instance.py
Comment (blackboxsw) [inline pycloudlib/ec2/instance.py:167]: While this feels strange to provide a separate associate_public_ip, the use-case and the calls to the botocore API are different given we are trying to associate a public IP address to the primary private_IP on this interface_id. This works for the use-cases we have today and provides us with a simple mechanism to assign a single public IP to a given interface without having to also having to creation secondary private IPs.

**Review by blackboxsw — APPROVED** [review]
It would be nice to add one example to examples/ec2.py (or adapt examples/ec2.py::hot_add) to create/associate a secondary IPv4 address) with an explanatory comment.

There are no unit_tests covering pycludlib/ec2/  at the moment. But, it doesn't have to block this PR at this time. It is just something we should start thinking about as we continue to evolve cloud-specific instance APIs to increase certainty that the python logic is correct and 'documented' in unit tests.

File: pycloudlib/ec2/instance.py
Comment (aciba90) [reply pycloudlib/ec2/instance.py]: Thanks for the simplification.

**Comment by aciba90** [comment]
Thanks, @blackboxsw, for the review and suggestions. I think I have addressed them all.

File: pycloudlib/azure/instance.py
Comment (blackboxsw) [inline pycloudlib/azure/instance.py:271]: Are we doing this twice? It looks like this is peformed 20 links below as well

**Review by blackboxsw — CHANGES_REQUESTED** [review]
I think there is a stray duplication of code in azure/instance.py. Once that is dropped, this is good to land.

File: pycloudlib/azure/instance.py
Comment (aciba90) [reply pycloudlib/azure/instance.py]: Wrong merge conflicts resolution, amended.

**Review by blackboxsw — APPROVED** [review]
Ship it!

**Pull request merged**
